mm2src/coins/rpc_command/init_scan_for_new_addresses.rs

```diff
@@ -157,6 +162,15 @@ pub mod common_impl {
             .scan_for_new_addresses(hd_wallet, hd_account.deref_mut(), &address_scanner, gap_limit)
             .await?;
 
+        let addresses: Vec<_> = new_addresses
+            .iter()
+            .map(|address_balance| Address::from_str(&address_balance.address).expect("Valid address"))
+            .collect();
+
+        prepare_addresses_for_balance_stream_if_enabled(coin, addresses)
+            .await
+            .map_err(|e| HDAccountBalanceRpcError::FailedScripthashSubscription(e.to_string()))?;
```

Comment:
As far as I see it gets even more complicated then.

e.g., getting known addresses number may instantly return an error for some protocols(not for now because it works for UTXO) according to: https://github.com/KomodoPlatform/komodo-defi-framework/blob/8a333a42e710d8e580fdcd898b483d1093e39446/mm2src/coins/hd_wallet.rs#L387-L389

Another problem is, calling `set_known_addresses_number` when `prepare_addresses_for_balance_stream_if_enabled` fails may also fail and in that case we will have the same problem.

Most efficient way of handling this that comes to my mind is changing the way hdwallet works by keeping these values in SQLite instead of memory, so we can insert all these values in transactional operation and revert it whenever we want (in this case it's when `prepare_addresses_for_balance_stream_if_enabled` fails). But it's very different piece of task to handle in this particular PR.

Reply:
> Most efficient way of handling this that comes to my mind is changing the way hdwallet works by keeping these values in SQLite instead of memory, so we can insert all these values in transactional operation and revert it whenever we want (in this case it's when prepare_addresses_for_balance_stream_if_enabled fails). But it's very different piece of task to handle in this particular PR.

We keep the values in sqlite, and we update the values in memory only when sqlite update has passed
https://github.com/KomodoPlatform/komodo-defi-framework/blob/cc3dfaca0f37967c4fc86d8359ceb35712a094fd/mm2src/coins/utxo/utxo_common.rs#L324-L335

> e.g., getting known addresses number may instantly return an error for some protocols(not for now because it works for UTXO) according to:

That's why it should be called before `scan_for_new_addresses` as I indicated here

> You would need to get both the internal and external known addresses number before this https://github.com/KomodoPlatform/komodo-defi-framework/blob/c3e264d19b37f198519d3f2296e9a869c4d5c2c0/mm2src/coins/rpc_command/init_scan_for_new_addresses.rs#L161-L163

In case it fails, there should be no problem since nothing has changed.

> Another problem is, calling set_known_addresses_number when prepare_addresses_for_balance_stream_if_enabled fails may also fail and in that case we will have the same problem.

Yep, but it can only fail due to DB operations failing. But an electrum request failing is far more common than DB failures, and if DB operations fail, it's a big problem in any part of our code anyways that I am sure we don't handle gracefully in other places.

> But it's very different piece of task to handle in this particular PR.

I agree that the best way is to call `set_known_addresses_number` only once at the end of the rpc call, which means removing the below and calling it in multiple places instead. But this is out of scope for this PR as you mentioned. I can do this in my HD wallet PRs after this is merged.
https://github.com/KomodoPlatform/komodo-defi-framework/blob/cc3dfaca0f37967c4fc86d8359ceb35712a094fd/mm2src/coins/utxo/utxo_common.rs#L450-L456

Reply:
>> e.g., getting known addresses number may instantly return an error for some protocols(not for now because it works for UTXO) according to:

> That's why it should be called before scan_for_new_addresses as I indicated here

I meant it will never work for that protocol then. I guess this is the expected behaviour?

> if DB operations fail, it's a big problem in any part of our code anyways that I am sure we don't handle gracefully in other places.

DB transactions. It's the only way of ensuring the data efficiency. Operation entrypoints should control the behaviour of db operations and should able to revert that in case of any failure.

Reply:
> I meant it will never work for that protocol then. I guess this is the expected behaviour?

Thanks you for pointing this out, `known_addresses_number` should be changed to not return an error at all since it will be a part of any HD wallet struct.

> DB transactions. It's the only way of ensuring the data efficiency. Operation entrypoints should control the behaviour of db operations and should able to revert that in case of any failure. Right now on any operation there can be left-over/dirty state on different parts of mm2 runtime and storage.

Totally agree.

Reply:
Refactored `prepare_addresses_for_balance_stream_if_enabled` as sending scripthash event instead of executing rpc. Now `prepare_addresses_for_balance_stream_if_enabled` can only fail when channel receiver is not available while trying to send message from sender, which should never happen (if there is no bug on the codebase).

Reply:
This is the best solution, thanks.